core/services/ocr2/plugins/ocr2keeper/evm21/transmit/cache.go

```diff
@@ -0,0 +1,80 @@
+package transmit
+
+import (
+	"encoding/hex"
+	"sync"
+
+	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg/v3/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+)
+
+// transmitEventCache holds a ring buffer of the last visited blocks (transmit block),
+// and their corresponding logs (by log id).
+// Using a ring buffer allows us to keep a cache of the last N blocks,
+// without having to iterate over the entire buffer to clean it up.
+type transmitEventCache struct {
+	lock   sync.RWMutex
+	buffer []cacheBlock
+
+	cap int64
+}
+
+func newTransmitEventCache(cap int64) transmitEventCache {
+	return transmitEventCache{
+		buffer: make([]cacheBlock, cap+1),
```

Comment:
nit. not sure why we need cap+1 here since when getting index we always do `int64(e.TransmitBlock) % int64(c.cap)` but no harm in keeping it as is